compiler/rustc_passes/src/check_attr.rs

```diff
@@ -1437,13 +1437,23 @@ impl<'tcx> CheckAttrVisitor<'tcx> {
                 | Target::Enum
                 | Target::Struct
                 | Target::Union
-                | Target::Method(_)
+                | Target::Method(MethodKind::Trait { body: false } | MethodKind::Inherent)
                 | Target::ForeignFn
                 // `impl Trait` in return position can trip
                 // `unused_must_use` if `Trait` is marked as
                 // `#[must_use]`
                 | Target::Trait
-        ) {
+        ) && !(matches!(
```

Comment:
Instead of two `matches!` conditions, maybe a match statement is now cleaner?

Reply:
After trying it, a `match` statement looked even uglier because of the additional guard on the second `matches!()`. But I've replaced this with two early returns which make the code easier to read IMO, and deindent the action part of the function. WDYT?

@rustbot review